File: csharp/GlobularServer/globularServer.cs
using System;
using System.IO;
using System.Text.Json;
using Grpc.Core;
using Grpc.Core.Interceptors;
using System.Threading.Tasks;

// TODO for the validation, use a map to store valid method/token/ressource/access
// the validation will be renew only if the token expire. And when a token expire
// the value in the map will be discard. That way it will put less charge on the server
// side.
namespace Globular
{

    /// <summary>
    /// That class contain the basic service class. Globular service are 
    /// plain gRPC service with required attributes to make it manageable.
    /// </summary>
    public class GlobularService
    {
        public string Name { get; set; }
        public int Port { get; set; }
        public int Proxy { get; set; }
        public string Protocol { get; set; }
        public bool AllowAllOrigins { get; set; }
        public string AllowedOrigins { get; set; }
        public string Domain { get; set; }
        public string CertAuthorityTrust { get; set; }
        public string CertFile { get; set; }
        public string KeyFile { get; set; }
        public bool TLS { get; set; }
        public string Version { get; set; }
        public string PublisherId { get; set; }
        public bool KeepUpToDate { get; set; }
        public bool KeepAlive { get; set; }

        private RessourceClient ressourceClient;
        public ServerUnaryInterceptor interceptor;

        /// <summary>
        /// The default constructor.
        /// </summary>
        public GlobularService(string address = "localhost")
        {
            // set default values.
            this.Domain = address;
            this.Protocol = "grpc";
            this.Version = "0.0.1";
            this.PublisherId = "localhost";

            // Create the interceptor.
            this.interceptor = new Globular.ServerUnaryInterceptor(this);
        }

        private RessourceClient getRessourceClient()
        {
            if (this.ressourceClient == null)
            {
                // there must be a globular server runing in order to validate ressources.
                ressourceClient = new RessourceClient(this.Domain, "Ressource");
            }
            return this.ressourceClient;
        }

        private string getPath()
        {
            return Directory.GetCurrentDirectory();
        }

        /// <summary>
        /// Initialyse from json object from a file.
        /// </summary>
        public object init(object server)
        {
            var configPath = this.getPath() + Path.DirectorySeparatorChar + "config.json";
            // Here I will read the file that contain the object.
            if (File.Exists(configPath))
            {
                var jsonStr = File.ReadAllText(configPath);
                var s = JsonSerializer.Deserialize(jsonStr, server.GetType());
                return s;
            }
            this.save(server);
            return server;
        }

        /// <summary>
        /// Serialyse the object into json and save it in config.json file.
        /// </summary>
        public void save(object server)
        {
            var configPath = getPath() + Path.DirectorySeparatorChar + "config.json";
            string jsonStr;
            jsonStr = JsonSerializer.Serialize(server);
            File.WriteAllText(configPath, jsonStr);
        }

        /// <summary>
        /// Set a ressource on the globular ressource manager.
        /// </summary>
        /// <param name="path">The path must begin by /. Like a unix file path</param>
        /// <param name="name">The name of the ressource must be unique in it contex (path + '/' + name)</param>
        /// <param name="modified">The last time the ressource was access</param>
        /// <param name="size">The size of the ressource (optional)</param>
        public void setRessource(string path, string name, int modified, int size)
        {
            this.getRessourceClient().SetRessource(path, name, modified, size);
        }

        /// <summary>
        /// Validate if a given user has write to use a given method
        /// </summary>
        /// <param name="token">Bearer Token</param>
        /// <param name="method"></param>
        /// <returns></returns>
        public bool validateUserAccess(string token, string method)
        {
            return this.getRessourceClient().ValidateApplicationAccess(token, method);
        }

        public bool validateApplicationAccess(string application, string method)
        {
            return this.getRessourceClient().ValidateApplicationAccess(application, method);
        }

        public bool validateUserRessourceAccess(string token, string path, string method, int permission)
        {
            return this.getRessourceClient().ValidateUserRessourceAccess(token, path, method, permission);
        }

        public bool validateApplicationRessourceAccess(string application, string path, string method, int permission)
        {
            return this.getRessourceClient().ValidateApplicationRessourceAccess(application, path, method, permission);
        }

        public int getActionPermission(string action)
        {
            return this.getRessourceClient().GetActionPermission(action);
        }

        public void logInfo(string application, string token, string method, string message, int logType)
        {
            this.getRessourceClient().Log(application, token, method, message, logType);
        }
    }

    public class ServerUnaryInterceptor : Interceptor
    {

        private GlobularService service;

        public ServerUnaryInterceptor(GlobularService srv)
        {
            this.service = srv;
        }

        public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(TRequest request, ServerCallContext context, UnaryServerMethod<TRequest, TResponse> continuation)
        {
            // Do method validations here.
            Metadata metadatas = context.RequestHeaders;
            string application = "";
            string token = "";
            string method = context.Method;
            bool hasAccess = false;
            // Console.Write("----> validate method: " + method);
            // Get the metadata from the header.
            for (var i = 0; i < metadatas.Count; i++)
            {
                var metadata = metadatas[i];
                if (metadata.Key == "application")
                {
                    application = metadata.Value;
                    if (!hasAccess)
                    {
                        hasAccess = this.service.validateApplicationAccess(application, method);
                    }
                }
                else if (metadata.Key == "token")
                {
                    token = metadata.Value;
                    if (!hasAccess)
                    {
                        hasAccess = this.service.validateUserAccess(token, method);
                    }
                }
            }

            // Here I will validate the user for action.
            if (!hasAccess)
            {
                // here I the user and the application has no access to the method 
                // I will throw an exception.
                throw new RpcException(new Status(StatusCode.PermissionDenied, "Permission denied"), metadatas);
            }

            // Now if the action has ressource access permission defines...
            var permission = this.service.getActionPermission(method);
            if (permission != -1)
            {
                // In that case a permission was set for the action so I will try to validate method parameters...

                // Now I will try to validate ressource if there is none...
                foreach (var prop in request.GetType().GetProperties())
                {
                    if (prop.PropertyType == typeof(string))
                    {
                        string path = (string)prop.GetValue(request, null);
                        if (path.StartsWith("/"))
                        {
                            var hasRessourcePermission = this.service.validateUserRessourceAccess(token, path, method, permission);
                            if(!hasRessourcePermission){
                                this.service.validateApplicationRessourceAccess(application, path, method, permission);
                            }
                            if(!hasRessourcePermission){
                                 throw new RpcException(new Status(StatusCode.PermissionDenied, "Permission denied"), metadatas);
                            }
                        }
                    }
                }
            }

            // this.service.
            var response = await base.UnaryServerHandler(request, context, continuation);
            return response;
        }
    }
}

File: cpp/GlobularServer/globularserver.cpp
#include "globularserver.h"
#include <string>
#include <fstream>
#include <iostream>
#include <sstream>
#include <map>
#include <cstddef>
#include <bitset>         // std::bitset
#include <math.h>       /* ceil */
#include "json.hpp"
#include <fstream>

#ifdef WIN32
#include <windows.h>
std::string getexepath()
{
    char result[MAX_PATH];
    return std::string(result, GetModuleFileNameA(NULL, result, MAX_PATH));
}

void sleep(unsigned milliseconds)
{
    Sleep(milliseconds);
}

#else
#include <limits.h>
#include <unistd.h>
#include <linux/limits.h>

std::string getexepath()
{
    char result[PATH_MAX];
    ssize_t count = readlink("/proc/self/exe", result, PATH_MAX);
    return std::string(result, (count > 0) ? count : 0);
}

#endif // WIN32

Globular::GlobularService::GlobularService(std::string id,
                                           std::string name,
                                           std::string domain,
                                           std::string publisher_id,
                                           bool allow_all_origins,
                                           std::string allowed_origins,
                                           bool tls,
                                           unsigned int defaultPort,
                                           unsigned int defaultProxy):
    id(id),
    name(name),
    domain(domain),
    publisher_id(publisher_id),
    allow_all_origins(allow_all_origins),
    port(defaultPort),
    proxy(defaultProxy),
    allowed_origins(allowed_origins),
    tls(tls)
{

    // first of all I will try to open the configuration from the file.
    std::ifstream inFile;
    std::string execPath = getexepath();
#ifdef WIN32
    std::size_t lastIndex = execPath.find_last_of("/\\");
    std::string configPath = execPath.substr(0, lastIndex) + "\\config.json";
#else
    std::size_t lastIndex = execPath.find_last_of("/");
    std::string configPath = execPath.substr(0, lastIndex) + "/config.json";
#endif
    inFile.open(configPath); //open the input file

    if (inFile.good()) {

        std::stringstream strStream;
        strStream << inFile.rdbuf(); //read the file
        std::string jsonStr = strStream.str(); //str holds the content of the file

        // Parse the json file.
        auto j = nlohmann::json::parse(jsonStr);

        // Now I will initialyse the value from the configuration file.
        this->publisher_id = j["PublisherId"];
        this->version = j["Version"];
        this->keep_up_to_date = j["KeepUpToDate"];
        this->allow_all_origins = j["AllowAllOrigins"];
        this->cert_authority_trust = j["CertAuthorityTrust"];
        this->keep_alive = j["KeepAlive"];
        this->cert_file = j["CertFile"];
        this->domain = j["Domain"];
        this->key_file = j["KeyFile"];
        this->name = j["Name"];
        this->port = j["Port"];
        this->proxy = j["Proxy"];
        this->path = j["Path"];
        this->proto = j["Proto"];
        this->tls = j["TLS"];
        this->protocol = j["Protocol"];

        // can be a list of string
        this->allowed_origins = j["AllowedOrigins"];
    }

    // Now I will create the new ressource client.
    this->ressourceClient = new RessourceClient("ressource.RessourceService", this->domain);
}

void Globular::GlobularService::save() {
    nlohmann::json j;
    j["PublisherId"] = this->publisher_id;
    j["Version"] = this->version;
    j["KeepUpToDate"] = this->keep_up_to_date;
    j["KeepAlive"] = this->keep_alive;
    j["AllowAllOrigins"] = this->allow_all_origins;
    j["AllowedOrigins"] = this->allowed_origins; // empty string
    j["CertAuthorityTrust"] = this->cert_authority_trust;
    j["CertFile"] = this->cert_file;
    j["Domain"] = this->domain;
    j["KeyFile"] = this->key_file;
    j["Name"] = this->name;
    j["Port"] = this->port;
    j["Id"] = this->id;
    j["Protocol"] = "grpc";
    j["Proto"] = this->proto;
    j["Proxy"] = this->proxy;
    j["TLS"] = this->tls;

    std::string execPath = getexepath();
    j["Path"] = execPath;

#ifdef WIN32
    std::size_t lastIndex = execPath.find_last_of("/\\");
    std::string configPath = execPath.substr(0, lastIndex) + "\\config.json";
#else
    std::size_t lastIndex = execPath.find_last_of("/");
    std::string configPath = execPath.substr(0, lastIndex) + "/config.json";
#endif
    std::ofstream file;
    file.open(configPath);
    file << j.dump();
    file.close();
}
